sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -259,11 +256,151 @@ class AzureAppConfigurationProvider(Mapping[str, Union[str, JSON]]):
     keys. Enables resolution of Key Vault references in configuration settings.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, **kwargs) -> None:
         self._dict: Dict[str, str] = {}
         self._trim_prefixes: List[str] = []
         self._client: Optional[AzureAppConfigurationClient] = None
         self._secret_clients: Dict[str, SecretClient] = {}
+        self._key_vault_options: Optional[
+            AzureAppConfigurationKeyVaultOptions
+        ] = kwargs.pop("key_vault_options", None)
+        self._refresh_options: Optional[
+            AzureAppConfigurationRefreshOptions
+        ] = kwargs.pop("refresh_options", AzureAppConfigurationRefreshOptions())
+        self._selects: List[SettingSelector] = kwargs.pop(
+            "selects", [SettingSelector(key_filter="*", label_filter=EMPTY_LABEL)]
+        )
+        self._sentinel_keys: List[str] = []
+
+        if self._refresh_options is not None:
+            self._next_refresh_time = datetime.now() + timedelta(
+                seconds=self._refresh_options._refresh_interval
+            )
+
+        trim_prefixes: List[str] = kwargs.pop("trim_prefixes", [])
+        self._trim_prefixes = sorted(trim_prefixes, key=len, reverse=True)
+        self._attempts = 1
+        self._min_backoff: Optional[int] = kwargs.pop("min_backoff", 30)
+        self._max_backoff: Optional[int] = kwargs.pop("min_backoff", 600)
+
+    def refresh(self) -> None:
+        if (
+            self._refresh_options is None
+            or len(self._refresh_options.get_refresh_registrations()) == 0
+        ):
+            return
+
+        if datetime.now() < self._next_refresh_time:
+            return
+
+        refresh_all = False
+
+        for registration in self._refresh_options.get_refresh_registrations():
+            if registration.refresh_all == True:
+                updated_sentinel = self._client.get_configuration_setting(
+                    key=registration.key_filter, label=registration.label_filter
+                )
+                if registration.etag == updated_sentinel.etag:
+                    continue
+                else:
+                    refresh_all = True
+                    break
+
+        if refresh_all:
```

Comment:
I would expect something like this to also be settable per method call.

Reply:
This is configured at startup if you look at the class `AzureAppConfigurationRefreshOptions` you can have keys that either refresh themselves or all configurations. This give more flexibility as the customer can decide at runtime how the refresh happens and what is refreshed.

We use keys as flags as we don't want a partial refresh. For example you have keys A, B, C, and D. If you are modifying all 4 keys you don't want a refresh to pull in just A and B while C and D are still the original values, we use a additional key to flag that all keys are safe to pull.

Reply:
Yes I saw that it was configured at start up, but usually where we have a client-level parameter, we provide the ability to overwrite that on request-by-request basis, unless it doesn't make sense to do so. Is it possible that I might want to change my refresh options on the fly without having to recreate a provider instance?

Reply:
So, the situation here is do all keys suddenly become refresh all keys? That seems confusing and misleading. And would seem like a possible force refresh which this isn't.

What situation would this be? They will still need to update their configurations and have some registered key to cause the refresh.

Also, this never recreates the provider, we are just updating the values of the stored keys. We store the client used connecting to the store, so this is just pulling in the updated keys.